File: BEMLIB_22.10/01_num_meth/06_interp_diff/splc_clm.f
      subroutine splc_clm
     +
     +  (N
     +  ,xp,fp
     +  ,slope1
     +  ,slope2
     +  ,a,b,c
     +  )

c-----------------------------------------
c FDLIB
c
c Copyright by C. Pozrikidis, 1999
c All rights reserved.
c
c This program is to be used only under the
c stipulations of the licensing agreement.
c----------------------------------------

c------------------------------------------------
c This program accompanies the book:
c
c           C. Pozrikidis
c ``Numerical Computation in Science and Engineering''
c      Oxford University Press
c------------------------------------------------

c--------------------------------------
c  Cubic-spline fit of prescribed data 
c  with clamped-end conditions
c
c  SYMBOLS
c  -------
c
c  N  .... number of intervals
c  xp .... x coord. of prescribed data
c  fp .... y coord. of prescribed data
c	
c  a .... polynomial coefficient related to 3rd derivative
c  b .... polynomial coefficient related to 2nd derivative
c  c .... polynomial coefficient related to 1st derivative
c  h .... interval between prescribed data
c
c  at.... diagonal of tridiagonal matrix
c  bt.... superdiagonal of tridiagonal matrix
c  ct.... subdiagonal of tridiagonal matrix
c
c CAPACITY:
c --------
c
c N=512 max
c--------------------------------------

      Implicit Double Precision (a-h,o-z)

      Dimension  xp(513),fp(513),b(513)
      Dimension   h(512)
      Dimension   a(512),c(512)

      Dimension at(512),bt(512),ct(512),rhs(512)
      Dimension        sln(512)

c--------
c prepare
c--------

      Na = N-1
      N1 = N+1

c-----------------------
c compute intervals h(i)
c-----------------------

      Do i=1,N
        h(i) = xp(i+1)-xp(i)
      End Do 

      h1 = h(1)
      hN = h(N)

c--------------------------------
C Generate the tridiagonal matrix
c--------------------------------
 
      at(1) = 2.0D0*(h(1)+h(2))- 0.50D0*h1         ! clamped end
      bt(1) = h(2)

      Do i=2,N-1
        i1 = i+1
        at(i) = 2.0D0*(h(i)+h(i1))
        bt(i) = h(i1)
        ct(i) = h(i)
      End Do 

      at(Na) = 2.0D0*(h(Na)+h(N)) - 0.50D0*hN         ! clamped end
      ct(Na) = h(Na)

      Do i=1,Na
        i1 = i+1
        i2 = i+2
        rhs(i)  = 3.0D0*( (fp(i2)-fp(i1))/h(i1)
     +                   -(fp(i1)-fp(i) )/h(i)  )
      End Do 

      rhs(1)  = rhs(1)  - 1.5D0*( (fp(2)  -fp(1))/h1 - slope1)  ! clamped end
      rhs(Na) = rhs(Na) + 1.5D0*( (fp(N+1)-fp(N))/hN - slope2)  ! clamped end
 
c------------------------
c solve the N-1 equations
c------------------------

      call thomas 
     +
     +   (Na
     +   ,at,bt,ct
     +   ,rhs
     +   ,sln
     +   )

c-----------------
c compute the bees
c-----------------

      Do i=1,Na
        b(i+1) = sln(i)
      End Do

      b(1)  = -0.5D0*b(2) + 1.5D0*( (fp(2) -fp(1))/h1 - slope1)/h1 ! clamped end
      b(N1) = -0.5D0*b(N) - 1.5D0*( (fp(N1)-fp(N))/hN - slope2)/hN ! clamped end

c---------------------------------
c compute the coefficients a and c
c---------------------------------

      Do i=1,N
        i1 = i+1
        a(i) = (b(i1)-b(i))/(3.0D0*h(i))
        c(i) = (fp(i1)-fp(i))/h(i) - h(i)*(b(i1)+2.0D0*b(i))/3.0D0
      End Do

c-----
c Done
c-----

      Return
      End

File: BEMLIB_22.10/01_num_meth/06_interp_diff/splc_clm.m
function [a,b,c] = splc_clm (N,x,f,slopeL,slopeR)

%-----------------------------------------
% FDLIB
%
% Copyright by C. Pozrikidis, 1999
% All rights reserved.
%
% This program is to be used only under the
% stipulations of the licensing agreement
%----------------------------------------

%------------------------------------------------
% This program accompanies the book:
%           C. Pozrikidis
% ``Numerical Computation in Science and Engineering''
%      Oxford University Press
%------------------------------------------------

%--------------------------------------
%  Cubic-spline fit of prescribed data
%  f(x) with clamped-end conditions
%
%  SYMBOLS
%  -------
%
%  N  .... number of intervals
%  x .... x coord. of prescribed data
%  f .... y coord. of prescribed data
%
%  a .... polynomial coefficient related to 3rd derivative
%  b .... polynomial coefficient related to 2nd derivative
%  c .... polynomial coefficient related to 1st derivative
%  h .... interval between prescribed data
%
%  a.... diagonal of tridiagonal matrix
%  b.... superdiagonal of tridiagonal matrix
%  c.... subdiagonal of tridiagonal matrix
%--------------------------------------

%-----------------------
% compute intervals h(i)
%-----------------------

  for i=1:N
   h(i) = x(i+1)-x(i);
  end

%--------------------------------
% generate a tridiagonal matrix
% for N-1 equations
%--------------------------------

  at(1) = 2.0D0*(h(1)+h(2))-0.50D0*h(1);
  bt(1) = h(2);

  for i=2:N-1
   at(i) = 2.0D0*(h(i)+h(i+1));
   bt(i) = h(i+1);
   ct(i) = h(i);
  end

  at(N-1) = 2.0D0*(h(N-1)+h(N))-0.50D0*h(N);
  ct(N-1) = h(N-1);

  for i=1:N-1
    rhs(i)  = 3.0D0*( (f(i+2)-f(i+1))/h(i+1) ...
                     -(f(i+1)-f(i) )/h(i)  );
  end

 rhs(1)   = rhs(1)   - 1.5*( (f(2)  -f(1))/h(1)-slopeL);
 rhs(N-1) = rhs(N-1) + 1.5*( (f(N+1)-f(N))/h(N)-slopeR);

%--------------------
% solve N-1 equations
%--------------------

  sln = thomas (N-1,at,bt,ct,rhs);

%-----------------
% recover the bees
%-----------------

  for i=1:N-1
   b(i+1) = sln(i);
  end

  b(1)  =-0.5D0*b(2)+1.5*((f(2)  -f(1))/h(1)-slopeL)/h(1);
  b(N+1)=-0.5D0*b(N)-1.5*((f(N+1)-f(N))/h(N)-slopeR)/h(N);

%---------------------------------
% compute the coefficients a and c
%---------------------------------

 for i=1:N
   a(i) = (b(i+1)-b(i))/(3.0D0*h(i));
   c(i) = (f(i+1)-f(i))/h(i)-h(i)*(b(i+1)+2.0D0*b(i))/3.0D0;
 end

%-----
% done
%-----

return
